### PR TITLE
Improved compatibility with Mac

### PR DESCRIPTION
**Pull request by mdevaev**
Using HID-Project on Mac was associated with two very big problems. This patch solves both - for the boot mouse and the absolute mouse. I believe these fixes could be relevant for other devices, but I have only dealt with these two.

1. When using any HID-Project mouse in Recovery Mode, Mac does can see it at the kernel level (that is, the cursor moves), but at the same time it asks connecting the mouse (which is already connected). The fact is that checking for the connected mouse is performed in a very stupid way: by searching for `USAGE(Pointer)` in the descriptors. Hardware mouse always (or almost always) have this, so adding this to the descriptor solves the problem. Details [here](https://github.com/pikvm/pikvm/issues/289) and [here](https://community.twocanoes.com/t/mds-automaton-keyboard-mouse-bluetooth-screen-in-macos-11-3/6391/3). I added this for absolute and boot mouses and it generally improves compatibility. It probably makes sense to do this for all mouses.

2. Another problem is that Apple UEFI (it uses another driver) could not work with BootMouse at all because it did not see it. It turned out that he needed a response to the `HID_HID_DESCRIPTOR_TYPE` request (which is strange, since we are already passing this descriptor in the interface). This is really part of the HID standard, and hardware mouse respond to this. Usually the BIOS and other things do not require this, but Apple UEFI turned out to be finicky.

I also added a bit of protocol and idle processing. It is better to let it be than not to be.

This patch is the result of my work with the hardware USB protocol analyzer. To understand the subtleties of Apple drivers, I compared how the Arduino behaves and how a "live" mouse works. All this is tested, of course.

### CODE REVIEW

**Comment by mdevaev** [comment]
Sorry for my terrible English, I'm very sleepy while writing this issue :)

File: src/SingleReport/BootMouse.cpp [hunk 1]
```diff
@@ -73,18 +78,24 @@ int BootMouse_::getInterface(uint8_t* interfaceCount)
...
-	if (setup.wValueH != HID_REPORT_DESCRIPTOR_TYPE) { return 0; }
-
 	// In a HID Class Descriptor wIndex cointains the interface number
 	if (setup.wIndex != pluggedInterface) { return 0; }
 
-	// Reset the protocol on reenumeration. Normally the host should not assume the state of the protocol
-	// due to the USB specs, but Windows and Linux just assumes its in report mode.
-	protocol = HID_REPORT_PROTOCOL;
+	// Check if this is a HID Class Descriptor request
+	if (setup.bmRequestType != REQUEST_DEVICETOHOST_STANDARD_INTERFACE) { return 0; }
```
Comment (NicoHood) [inline src/SingleReport/BootMouse.cpp:85]: Why did you move this down?

File: src/SingleReport/BootMouse.cpp [hunk 2]
```diff
@@ -73,18 +78,24 @@ int BootMouse_::getInterface(uint8_t* interfaceCount)
...
+
+	if (setup.wValueH == HID_HID_DESCRIPTOR_TYPE) {
+		// Apple UEFI wants it
+		HIDDescDescriptor desc = D_HIDREPORT(sizeof(_hidReportDescriptorMouse));
+		return USB_SendControl(0, &desc, sizeof(desc));
```
Comment (NicoHood) [inline src/SingleReport/BootMouse.cpp:90]: Do we need this for the bootkeyboard as well?

**Comment by NicoHood** [comment]
Thanks a lot for sending this in!

It looks like you've got very good equipment and knowledge about USB (as much knowledge as you can have with that very weird spec + try &error). Would you mind having a look at these issue with the bootkeyboard? Maybe you have also a device that does not work, and the knowledge to fix that as well?

https://github.com/NicoHood/HID/issues/34
https://github.com/NicoHood/HID/issues/225

File: src/SingleReport/BootMouse.cpp (hunk 1 above)
Comment (mdevaev) [reply src/SingleReport/BootMouse.cpp:85]: To make if-else branch for `setup.wValueH`. Previously, there was an inequality and a return, and now there is a check for two types of descriptor.

File: src/SingleReport/BootMouse.cpp (hunk 2 above)
Comment (mdevaev) [reply src/SingleReport/BootMouse.cpp:90]: Totally yes. This is the second problem that I described in the issue. This is how hardware mice behave, I also found it in the datasheet for some [mouse chip](https://www.silabs.com/documents/public/application-notes/AN249.pdf). Almost all the devices I have personally met do not require this command, but all Macs do. It's funny that this constant exists in the HID library, but is not used. Apparently, no one thought that the command associated with it could be mandatory.

**Comment by mdevaev** [comment]
To solve these problems, I need samples of problematic devices. Everything works perfectly on all the motherboards I have. I am aware of the problems with KVM, but not with the BIOS. Your library is used by many thousands of people through my project, and none of them said that there are problems with the BIOS and keyboard in the Arduino. On the contrary, I recommend Arduino for crappy BIOSes if the Linux Gadget does not work well.

I'll keep you posted. I'm even interested in finding out what's wrong there myself. If I get my hands on a broken device, I will investigate it and let you know. If you need anything, you can write to me by email mdevaev@gmail.com or discord https://discord.gg/bpmXfz5

**Comment by NicoHood** [comment]
What is the project you are working on?

Did you notice my comments to the code?

**Comment by mdevaev** [comment]
I'm doing [this](https://github.com/pikvm/pikvm), and [hardware for it](https://www.youtube.com/watch?v=dTchVKxx7Fo).

I responded to your comments in the review. Don't you see them?

File: src/SingleReport/BootMouse.cpp (hunk 1 above)
Comment (NicoHood) [reply src/SingleReport/BootMouse.cpp:85]: Yes I saw this, but this check can be kept above as well, right? I mean it should not matter and I also do not prefer anything. It is just for my understanding, that this line could also be placed above the wIndex check, right?

File: src/SingleReport/BootMouse.cpp (hunk 2 above)
Comment (NicoHood) [reply src/SingleReport/BootMouse.cpp:90]: So I guess you will be trying to fix the keyboard in another PR? I think this PR is ready to merge then, I dont see anything critical that should not work.

File: src/SingleReport/BootMouse.cpp (hunk 1 above)
Comment (mdevaev) [reply src/SingleReport/BootMouse.cpp:85]: Yes, there is no difference. I put the index above for reasons that first you need to check whether there was a request to this device, and then proceed to check the type of request. It would be strange to check the type first, then the index, and then the other type.

File: src/SingleReport/BootMouse.cpp (hunk 2 above)
Comment (mdevaev) [reply src/SingleReport/BootMouse.cpp:90]: Oh, crap, I'm sorry, I missed that you asked about the keyboard, and continued about the mouse. Stupid me. No, apparently this is not necessary for the keyboard. At least Apple and everything else works. But, POTENTIALLY, this may be due to that bug with KVM and strange BIOSes. Maybe if you add this code, everything will work. In the analysis of USB data, I have not seen a reference to this command, and I have not yet been able to find a dataset for some keyboard chip where this would be described. I would ask someone to experiment, but it's definitely not worth including this for the keyboard until we're sure.

**Comment by Bugerman21** [comment]
Hi All,
One year ago I have used this library on Leonardo, to simulate keystrokes on the keyboard from windows, boot menu, and BIOS, and check Caps Lock led status.
Now, I flow a similar problem with a new BIOS on Dell laptops.
When I am entering the boot menu or the BIOS, the laptop did not get from Leonardo a keystrokes.

Maybe this will help to someone for investigating the problem.
**At this moment I found this solution that works with UNO:**
I am sending commands from Leonardo to UNO.
https://www.instructables.com/How-to-Make-a-Arduino-HID-Keyboard/

**Comment by NicoHood** [comment]
This would possibly be the better solution:
https://github.com/NicoHood/HID/issues/225
